merge.cpp

```diff
@@ -61,6 +193,160 @@ void GetSnapshotDim(const int id, const std::string& basename, const std::string
+        // which include the last overlapping snapshot in previous time window, all the snapshots taken strictly before windowRight, 
+        // and the overlapping snapshot just taken at or after windowRight, making sure no data is missed by closing the basis window at or before windowRight
+        for (int paramID = 0; paramID < nset; ++paramID)
+        {
+            for (int t = 0; t < windowNumSamples + 2; ++t)
+            {
+                if (tSnapX[paramID].back() < windowRight)
+                {
+                    tSnapX[paramID].pop_back();
+                    offsetCurrentWindow[paramID+nset*VariableName::X] += 1;
+                }
+                if (tSnapV[paramID].back() < windowRight)
+                {
+                    tSnapV[paramID].pop_back();
+                    offsetCurrentWindow[paramID+nset*VariableName::V] += 1;
+                }
+                if (tSnapE[paramID].back() < windowRight)
+                {
+                    tSnapE[paramID].pop_back();
+                    offsetCurrentWindow[paramID+nset*VariableName::E] += 1;
+                }
+
+                if (rhsBasis)
+                {
+                    if (tSnapFv[paramID].back() < windowRight)
+                    {
+                        tSnapFv[paramID].pop_back();
+                        offsetCurrentWindow[paramID+nset*VariableName::Fv] += 1;
+                    }
+                    if (tSnapFe[paramID].back() < windowRight)
+                    {
+                        tSnapFe[paramID].pop_back();
+                        offsetCurrentWindow[paramID+nset*VariableName::Fe] += 1;
+                    }
+                }
+            }
+        }
+        offsetAllWindows.push_back(offsetCurrentWindow);
+        numBasisWindows += 1;
+
+        // Find the largest time, windowLeft, such that the last snapshot is counted for every variable and parameter
+        // The next basis window takes this snapshot, making sure no data is missed by opening the next basis window at or after windowLeft 
+        // By opening new basis window at the midpoint of windowLeft and windowRight, 
+        // we make sure no data is missed in both basis windows 
+        for (int paramID = 0; paramID < nset; ++paramID)
+        {
+            tTemp[paramID+nset*VariableName::X] = tSnapX[paramID].back();
+            tTemp[paramID+nset*VariableName::V] = tSnapV[paramID].back();
+            tTemp[paramID+nset*VariableName::E] = tSnapE[paramID].back();
+
+            if (rhsBasis)
+            {
+                tTemp[paramID+nset*VariableName::Fv] = tSnapFv[paramID].back();
+                tTemp[paramID+nset*VariableName::Fe] = tSnapFe[paramID].back();
+            }
+        }
+
+        double windowLeft = *max_element(tTemp.begin(), tTemp.end());
+        double overlapMidpoint = (windowLeft + windowRight) / 2;
+        twepTemp.push_back(overlapMidpoint);
+
+        if (numSnap[0] == offsetCurrentWindow[0]+1)
```

Comment:
This checks only variable `X`. Should all variables be checked to see if their samples have been used up, in which case this should be the final window?